ecsmetadata/client.go

```diff
@@ -18,44 +18,48 @@ package ecsmetadata
 import (
 	"context"
 	"encoding/json"
+	"fmt"
 	"io/ioutil"
-	"log"
 	"net/http"
 	"os"
 )
 
-var endpoint string
+type Client struct {
+	// HTTClient is the client to use when making HTTP requests when set.
+	HTTPClient *http.Client
 
-func init() {
-	const endpointEnv = "ECS_CONTAINER_METADATA_URI_V4"
-	endpoint = os.Getenv(endpointEnv)
-	if endpoint == "" {
-		log.Fatalf("%q environmental variable is not set, are you running this on ECS?", endpointEnv)
-	}
+	// metadata server endpoint
+	endpoint string
 }
 
-type Client struct {
-	// httpClient is the client to use when making HTTP requests
-	// when set. Otherwise, an internal default client will be used.
-	httpClient *http.Client
+// NewClient returns a new Client. endpoint is the metadata server endpoint.
+func NewClient(endpoint string) *Client {
+	return &Client{
+		HTTPClient: &http.Client{},
+		endpoint:   endpoint,
+	}
 }
 
-func NewClient(httpClient *http.Client) (*Client, error) {
-	if httpClient == nil {
-		httpClient = &http.Client{}
+// NewClientFromEnvironment is like NewClient but endpoint
+// is discovered from the environment.
+func NewClientFromEnvironment() (*Client, error) {
+	const endpointEnv = "ECS_CONTAINER_METADATA_URI_V4"
+	endpoint := os.Getenv(endpointEnv)
```

Comment:
I wonder if we should do a test request here to validate the URL in the ENV.

Reply:
I would not do a test request. But we could log the URL and attempt to parse it.

Reply:
Yea, that makes sense. `url.Parse()` would be good.